certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+
+    assert cancelerChanged && nonClaimRoot => !otherCancelerChanged;
+    assert revokerChanged && nonClaimRoot => !otherRevokerChanged;
+    assert granterChanged && nonClaimRoot => !otherGranterChanged;
+}
+
+
+// STATUS - verified
+// When `_root` is changed, it is only by calling `claimRoot`,
+// the caller must be the `_pendingRoot` before the call
+// and the new `_root` must also be the `_pendingRoot` before the call.
+// and variables are updated appropriately.
+// https://prover.certora.com/output/40577/0a8cadb54810435c859098750efa8ee0/?anonymousKey=58e33df6c4063dd4882ec3f40d8b86d816951102
+rule rootChangesOnlyWithClaimRoot(env e, method f) {
+    address rootBefore = getRoot();
+    address pendingRootBefore = getPendingRoot();
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    address rootAfter = getRoot();
+
+    // if the function changed the root, the sender was pendingRoot
+    assert rootBefore != rootAfter =>
+        e.msg.sender == pendingRootBefore,
+        "Root changed by somebody, who was not pending root.";
+    // if the function changed the root, the new root is old pendingRoot
+    assert rootBefore != rootAfter =>
+        rootAfter == pendingRootBefore,
+        "Pending root changed root to somebody else.";
+    // if the function f changed the root, then f must be claimRoot
+    assert rootBefore != rootAfter =>
+        f.selector == sig:claimRoot().selector,
+        "Root changed by a function other than claimRoot.";
+}
+
+
+// STATUS - verified
+// Only two functions can change pending root: claimRoot and setPendingRoot.
+// claimRoot can only be called by the execution helper.
+// setPendingRoot can only be called by the root or the execution helper.
+// https://vaas-stg.certora.com/output/40577/7efc90865f0f47d09d3d42dab4214400/?anonymousKey=26431befd59fbb32ae6201c42e7f456e7ccceb06
+rule pendingRootChangesOnlyWithSetPendingRootOrClaimRoot(env e, method f) {
+    address pendingRootBefore = getPendingRoot();
+    address executionHelper = getTimelockExecutionHelper();
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    address pendingRootAfter = getPendingRoot();
+
+    // if the function f changed the root, then f must be setPendingRoot or claimRoot.
+    assert pendingRootBefore != pendingRootAfter =>
+        ( f.selector == sig:setPendingRoot(address).selector ||
+        f.selector == sig:claimRoot().selector ),
+        "Pending root changed by a function other than setPendingRoot or claimRoot.";
+    // if the function changed the pending root, the sender was pending root or execution helper.
+    assert pendingRootBefore != pendingRootAfter =>
+        e.msg.sender == pendingRootBefore ||
+        e.msg.sender == executionHelper,
+        "Pending root changed by somebody, who was not root.";
+    assert (pendingRootBefore == 0 && pendingRootAfter != 0) => e.msg.sender == executionHelper,
+        "pendingRoot can only be changed by the executer";
```

Comment:
What's the difference between these asserts?

Reply:
The latter one has stronger assumption and so it can have stronger conclusion. I will modify it for better readability.

Reply:
one of them can be removed then, I guess

Reply:
or combined into one

Reply:
Well one does not imply the other. So there is no reason to remove one of these.
What would be an elegant way to combine these two assertion into one assertion, that would be equivalent to the two?